tests/std/tests/GH_002488_promise_not_default_constructible_types/test.cpp

```diff
@@ -0,0 +1,184 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+
+#include <cassert>
+#include <future>
+#include <thread>
+
+
+static std::atomic<int> has_default_objects{0};
+static std::atomic<int> no_default_objects{0};
+static std::atomic<int> no_default_or_assign_objects{0};
+
+
+struct has_default {
+    has_default() : x(0xbad) {
+        ++has_default_objects;
+    }
+    explicit has_default(int n) : x(n) {
+        ++has_default_objects;
+    }
+    has_default(const has_default& v) : x(v.x) {
+        ++has_default_objects;
+    }
+
+    ~has_default() {
+        --has_default_objects;
+    }
+
+    int x;
+};
+
+struct no_default {
+    no_default() = delete;
+    explicit no_default(int n) : x(n) {
+        ++no_default_objects;
+    }
+    no_default(const no_default& v) : x(v.x) {
+        ++no_default_objects;
+    }
+
+    ~no_default() {
+        --no_default_objects;
+    }
+
+    int x;
+};
+
+struct no_default_or_assign {
+    no_default_or_assign() = delete;
+    explicit no_default_or_assign(int n) : x(n) {
+        ++no_default_or_assign_objects;
+    }
+    no_default_or_assign(const no_default_or_assign& v) : x(v.x) {
+        ++no_default_or_assign_objects;
+    }
+
+    void operator=(const no_default_or_assign&) = delete;
+
+    ~no_default_or_assign() {
+        --no_default_or_assign_objects;
+    }
+
+    int x;
+};
+
+template <class F>
+void assert_throws_future_error(F f, std::error_code errc) {

```

Comment:
Naming suggestion: Although this isn't ambiguous/shadowing (because the test isn't `using namespace std;`), I recommend against choosing Standard identifiers for names in test code. Such avoidance increases readability by making Standard machinery be instantly recognizable (i.e. being able to assume that an `errc` refers to `std::errc`, especially because many of our tests are `using namespace std;`).

Here, `ec` or `code` would avoid the naming pseudo-collision.